server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java

```diff
@@ -0,0 +1,215 @@
+      metricTemplate.setDomain(domain);
+    }
+    if (description != null) {
+      metricTemplate.setDescription(description);
+    }
+    return metricTemplate;
+  }
+
+  private void updateMetricInDb(DbSession dbSession, MetricDto metricInDb, MetricDto metricTemplate) {
+    String key = metricTemplate.getKey();
+    String name = metricTemplate.getShortName();
+    String type = metricTemplate.getValueType();
+    String domain = metricTemplate.getDomain();
+    String description = metricTemplate.getDescription();
+    if (key != null) {
+      metricInDb.setKey(key);
+    }
+    if (name != null) {
+      metricInDb.setShortName(name);
+    }
+    if (type != null) {
+      metricInDb.setValueType(type);
+    }
+    if (domain != null) {
+      metricInDb.setDomain(domain);
+    }
+    if (description != null) {
+      metricInDb.setDescription(description);
+    }
+    dbClient.metricDao().update(dbSession, metricInDb);
+    dbSession.commit();
+  }
+
+  private void checkMetricInDbAndTemplate(DbSession dbSession, @Nullable MetricDto metricInDb, MetricDto template) {
+    if (isMetricFoundInDb(metricInDb) || isMetricDisabled(metricInDb) || isMetricNonCustom(metricInDb)) {
+      throw new ServerException(HttpURLConnection.HTTP_CONFLICT, String.format("No active custom metric has been found for id '%d'. Maybe you want to create a metric ?",
+        template.getId()));
+    }
+    if (haveMetricTypeChanged(metricInDb, template)) {
+      List<CustomMeasureDto> customMeasures = dbClient.customMeasureDao().selectByMetricId(dbSession, metricInDb.getId());
+      if (haveAssociatedCustomMeasures(customMeasures)) {
+        throw new ServerException(HttpURLConnection.HTTP_CONFLICT, String.format("You're trying to change the type '%s' while there are associated custom measures.",
+          metricInDb.getValueType()));
+      }
+    }
+  }
+
+  private void writeMetric(JsonWriter json, MetricDto metric) {
+    json.beginObject();
+    json.prop(FIELD_ID, String.valueOf(metric.getId()));
+    json.prop(FIELD_KEY, metric.getKey());
+    json.prop(FIELD_TYPE, metric.getValueType());
+    json.prop(FIELD_NAME, metric.getShortName());
+    json.prop(FIELD_DOMAIN, metric.getDomain());
+    json.prop(FIELD_DESCRIPTION, metric.getDescription());
+    json.endObject();
+  }
+
+  private boolean isMetricNonCustom(@Nullable MetricDto metricInDb) {
+    return !metricInDb.isUserManaged();
+  }
+
+  private boolean isMetricDisabled(@Nullable MetricDto metricInDb) {
+    return !metricInDb.isEnabled();
+  }
+
+  private boolean isMetricFoundInDb(@Nullable MetricDto metricInDb) {
+    return metricInDb == null;
+  }
+
+  private boolean haveAssociatedCustomMeasures(List<CustomMeasureDto> customMeasures) {
+    return !customMeasures.isEmpty();
+  }
+
+  private boolean haveMetricTypeChanged(@Nullable MetricDto metricInDb, MetricDto template) {
+    return !metricInDb.getValueType().equals(template.getValueType()) && template.getValueType() != null;
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) metricInDb must be non-null but is marked as nullable [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)